### PR TITLE
Add TCP_KEEPALIVE

### PR DESCRIPTION
**Pull request by iksaif**
The RATIO_RESET settings are not always robust enough to detect
all TCP issues. Also add the possibility to enable TCP KeepAlive
which is design to detect broken connections.

### CODE REVIEW

File: lib/carbon/util.py [hunk 1]
```diff
@@ -33,6 +33,11 @@
   import pickle
   USING_CPICKLE = False
 
+import socket
+from time import sleep, time
```
Comment (deniszh) [inline lib/carbon/util.py]: time already imported

File: lib/carbon/util.py [hunk 2]
```diff
@@ -33,6 +33,11 @@
...
+from twisted.python.util import initgroups
```
Comment (deniszh) [inline lib/carbon/util.py]: already imported in line 15

File: lib/carbon/util.py [hunk 3]
```diff
@@ -33,6 +33,11 @@
...
+from twisted.scripts.twistd import runApp
```
Comment (deniszh) [inline lib/carbon/util.py]: same here

**Comment by deniszh** [comment]
Lint is also very angry

File: lib/carbon/util.py [hunk 4]
```diff
@@ -42,6 +43,25 @@ def dropprivs(user):
   return (uid, gid)
 
 
+def enableTcpKeepAlive(transport):
+  from carbon.conf import settings
+
+  if not settings.TCP_KEEPALIVE or not hasattr(transport, 'getHandle'):
```
Comment (DanCech) [inline lib/carbon/util.py]: I'd prefer to see the setting passed in when the function is called, that way it's clearer what's going on.

Comment (iksaif) [reply lib/carbon/util.py]: or can I name it `maybeEnableTcpKeepAlive()` and pass it `settings` ? Less code duplication

Comment (DanCech) [reply lib/carbon/util.py]: I would prefer to just pass the setting itself, not a big fan of the `maybe` pattern

Comment (iksaif) [reply lib/carbon/util.py]: done

**Comment by DanCech** [comment]
Apologies, I didn't see that settings is also used for KEEPINT etc, in that case it'll be cleaner to pass in settings directly I guess.

**Comment by iksaif** [comment]
The new version takes 'enabled' + additional settings, so it's slightly better than better.

**Pull request merged**